app/src/main/java/com/unc0ded/shopdeliver/views/fragments/LoginFragment.java

```diff
@@ -0,0 +1,276 @@
+                        else
+                            Toast.makeText(getContext(), "No internet connection!", Toast.LENGTH_LONG).show();
+                        break;
+                    }
+                }
+            }
+        });
+
+        mCallbacks = new PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
+            @Override
+            public void onVerificationCompleted(@NonNull PhoneAuthCredential phoneAuthCredential) {
+                Log.d("customerCallbacks", "onVerificationCompleted:" + phoneAuthCredential);
+                binding.progressBar.setVisibility(View.GONE);
+                //signInWithPhoneAuthCredentials(phoneAuthCredential);
+            }
+
+            @Override
+            public void onVerificationFailed(@NonNull FirebaseException e) {
+                binding.progressBar.setVisibility(View.GONE);
+                Log.d("customerCallbacks", "onVerificationFailed", e);
+                Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
+                binding.emailOrPhone.setEnabled(true);
+                if (e instanceof FirebaseAuthInvalidCredentialsException) {
+                    Log.d("customerCallbacks","onVerificationFailed: Invalid Number");
+                }
+                else if (e instanceof FirebaseTooManyRequestsException) {
+                    Log.d("customerCallbacks","onVerificationFailed: SMS quota exceeded");
+                }
+            }
+
+            @Override
+            public void onCodeSent(@NonNull String s, @NonNull PhoneAuthProvider.ForceResendingToken forceResendingToken) {
+                binding.progressBar.setVisibility(View.GONE);
+                super.onCodeSent(s, forceResendingToken);
+                verificationId = s;
+
+                Toast.makeText(requireContext(), "OTP has been sent", Toast.LENGTH_SHORT).show();
+
+                OtpWidget otpView = new OtpWidget(requireContext());
+                otpView.getVerifyBtn().setOnClickListener(view -> {
+                    if (Objects.requireNonNull(otpView.getOtpView().getText()).toString().length() == 6) {
+                        PhoneAuthCredential customerCredential = PhoneAuthProvider.getCredential(verificationId, otpView.getOtpView().getText().toString());
+                        loginActivityVM.signInWithPhone(customerCredential);
+                        enter_otp.hide();
+                    }
+                    else
+                        Toast.makeText(requireContext(), "Please enter complete OTP", Toast.LENGTH_SHORT).show();
+                });
+                enter_otp = new MaterialAlertDialogBuilder(requireContext())
+                        .setTitle("Enter OTP")
+                        .setCancelable(true)
+                        .setView(otpView)
+                        .show();
+            }
+        };
+    }
```

Comment:
Can't this whole part go into the Repository too? As for the UI tasks, those can be done using LiveData flags/listeners I guess.